### PR TITLE
Show return type hints on last line of pipeline chain

### PR DESCRIPTION
**Pull request by DedSec256**
https://youtrack.jetbrains.com/issue/RIDER-77585/Show-return-type-hints-on-last-line-of-pipeline-chain

image

### CODE REVIEW

**Comment by DedSec256** [comment]
image
image
image

**Comment by auduchinok** [comment]
@DedSec256 Please rebase on `main`.

File: ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Error 03 - Without right argument.fs.gold [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿[ 1 ]||(0)
+|>
```
Comment (auduchinok) [inline ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Error 03 - Without right argument.fs.gold:2]: This is a good one!

File: ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Multi line 03 - Same line.fs [hunk 2]
```diff
@@ -7,4 +7,4 @@
 |> Set.ofSeq |>
 Set.toSeq
 |> Set.ofSeq
-|> ignore
+|> id
```
Comment (auduchinok) [inline ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Multi line 03 - Same line.fs:10]: 👍

File: ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Skipped 04 - Unit.fs.gold [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+|> ignore
+
+[1, 2]||(5)
+|> dict||(6)
+
+[1]||(7)
+|> List.iter ignore
+
+[1]||(8)
+|> List.map ignore||(9)
```
Comment (auduchinok) [inline ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Skipped 04 - Unit.fs.gold:17]: 👍

File: ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Skipped 04 - Unit.fs.gold [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+5||(10)
+|> Some||(11)
+
+
+
+let f _ _ = ()
+
+1||(12)
+|> f||(13)
```
Comment (auduchinok) [inline ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Skipped 04 - Unit.fs.gold:27]: 👍

File: ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Skipped 05 - Inner pipe.fs.gold [hunk 5]
```diff
@@ -0,0 +1,99 @@
+﻿if true then
+      1||(0)
+      |> id
+else 1||(1)
+     |> id||(2)
```
Comment (auduchinok) [inline ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Skipped 05 - Inner pipe.fs.gold:5]: Is it expected that one branch shows the type and the other does not? Is it to skip the duplicated type?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PipeChainTypeHintStage.fs [hunk 6]
```diff
@@ -24,22 +28,54 @@ type SameLinePipeHints =
...
+            context.Add(opExpr, leftArgument, true)
+
+        if isNull rightArgument || rightArgument :? IFromErrorExpr then () else
+
+        let isTopBinary =
+            let outermostParent =
+                binaryAppExpr
+                 .IgnoreParentParens()
+                 .GetOutermostParentExpressionFromItsReturn()
+                 .IgnoreParentParens()
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PipeChainTypeHintStage.fs:61]: I wonder if this parens skipping would be better to be inside `GetOutermostParentExpressionFromItsReturn` instead of having to do it twice later?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PipeChainTypeHintStage.fs [hunk 7]
```diff
@@ -54,19 +90,46 @@ type PipeOperatorVisitor(sameLinePipeHints: SameLinePipeHints) =
 type PipeChainHighlightingProcess(logger: ILogger, fsFile, settings: IContextBoundSettingsStore, daemonProcess: IDaemonProcess) =
     inherit FSharpDaemonStageProcessBase(fsFile, daemonProcess)
 
-    let adornExprs logKey (exprs : (IReferenceExpr * ITreeNode)[]) =
+    let isRightPipeArgApplicable (expr: ITreeNode) (pipeResultType: FSharpType) =
+        if not (isUnit pipeResultType) then true else
+
+        let rec skipFunArgs (t: FSharpType) =
+            if t.IsFunctionType then skipFunArgs t.GenericArguments[1] else t
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PipeChainTypeHintStage.fs:97]: I'm not sure I can easily follow the logic here. Could you add some comments/examples, please?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PipeChainTypeHintStage.fs [hunk 8]
```diff
@@ -54,19 +90,46 @@ type PipeOperatorVisitor(sameLinePipeHints: SameLinePipeHints) =
...
             if isNull symbolUse then () else
 
-            let _, fcsType = symbolUse.GenericArguments[0]
+            let _, fcsType = symbolUse.GenericArguments[if isLeft then 0 else 1]
+            let isApplicable = isLeft || isRightPipeArgApplicable exprToAdorn fcsType
+
+            if not isApplicable then () else
             let range = exprToAdorn.GetNavigationRange().EndOffsetRange()
-            highlightingConsumer.AddHighlighting(TypeHintHighlighting(fcsType.Format(symbolUse.DisplayContext), range))
+            let displayContext = symbolUse.DisplayContext.WithShortTypeNames(true)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PipeChainTypeHintStage.fs:131]: Could be nice to be in a separate PR, but OK. Could you add some tests for this change, please?

**Review by auduchinok — COMMENTED** [review]
Very good testing overall, but I'm a bit lost with the `skipFunArgs` helper.

File: ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Skipped 05 - Inner pipe.fs.gold (hunk 5 above)
Comment (DedSec256) [reply ReSharper.FSharp/test/data/features/daemon/pipeChainTypeHints/Skipped 05 - Inner pipe.fs.gold:5]: actually, after 
```F#
else 1
     |> id
``` 
hint is shown for the whole if/else expression as a left operand of outer pipe. For the `else` branch, we are indeed not displaying the duplicated type (even if it may be slightly different). For the `then` branch and in the previous branches in the `match`, I have decided not to display the type, as we will show the type for the entire if/match below.

In C#, Linq similarly shows the type for inner Linq only on the last branch for switch expression
image

**Review by auduchinok — APPROVED** [review]
